File: C++_Version/Linked List/203Remove_Linked_List_Elements.cpp
//Remove all elements from a linked list of integers that have value val.
//
//Example
//Given : 1 --> 2 --> 6 --> 3 --> 4 --> 5 --> 6, val = 6
//	Return: 1 --> 2 --> 3 --> 4 --> 5

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
using namespace std;


//* Definition for singly-linked list.
struct ListNode 
{
	int val;
	ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* removeElements(ListNode* head, int val) {
		ListNode *ptemp = head,*pdelete;

		while (true)                             //保证处理后头结点的值不是val
		{
			if (ptemp == NULL)
				return NULL;

			if (ptemp->val == val)
			{
				pdelete = ptemp;
				ptemp = ptemp->next;
				delete pdelete;
			}
			else
			{
				head = ptemp;
				break;
			}
		}

		while (true)
		{
			if (ptemp->next == NULL)
				return head;
			else
			{
				if (ptemp->next->val == val)
				{
					pdelete = ptemp->next;
					ptemp->next = ptemp->next->next;
					delete pdelete;
				}
				else
					ptemp = ptemp->next;
			}	
		}
	}
};

int main()
{
	ListNode* headA = new ListNode(0), *ptemp1, *ptemp2 = headA ;

	int a[8] = { 0,1,2,6,3,4,5,6 };

	for (int i = 1; i < 8; ++i)
	{
		ptemp1 = new ListNode(a[i]);
		ptemp2->next = ptemp1;
		ptemp2 = ptemp1;
	}


	Solution sol1;
	ListNode* receive = sol1.removeElements(headA, 6);

	system("pause");

	return 0;
}

File: Java_Version/203Remove_Linked_List_Elements.java
/* Remove all elements from a linked list of integers that have value val.

Example
Given: 1 --> 2 --> 6 --> 3 --> 4 --> 5 --> 6, val = 6
Return: 1 --> 2 --> 3 --> 4 --> 5

Credits:
Special thanks to @mithmatt for adding this problem and creating all test cases. */


import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;


public class Main {

    public static void main(String[] args) throws ArithmeticException {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(2);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(4);
        root.right.left = new TreeNode(4);
        root.right.right = new TreeNode(3);

        boolean result = new Solution().isSymmetric(root);

        System.out.println(result);
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {
    public ListNode removeElements(ListNode head, int val) {

        if (head == null) {
            return head;
        }

        ListNode slowPointer = head;
        while (slowPointer != null && slowPointer.val == val) {
            slowPointer = slowPointer.next;
        }
        if (slowPointer == null || slowPointer.next == null) {
            return slowPointer;
        }

        ListNode newHead = slowPointer;
        ListNode fastPointer = slowPointer.next;

        while (fastPointer != null) {

            if (fastPointer.val != val) {
                slowPointer = fastPointer;
                fastPointer = fastPointer.next;
            } else {
                while (fastPointer != null && fastPointer.val == val) {
                    fastPointer = fastPointer.next;
                }
                slowPointer.next = fastPointer;
                if (fastPointer == null) {
                    return newHead;
                } else {
                    slowPointer = fastPointer;
                    fastPointer = fastPointer.next;
                }
            }

        }

        return newHead;
    }
}